### PR TITLE
multiple improvements

### LINKED ISSUE #19: VGM complaining about task running too long
Hi all,

We're seeing an issue in one of our environments where we're seeing occasional failures of VGM to provide a token, with the error message: 

`2016/10/20 18:30:39 Failed to create token pair for 10.3.1.197:58262 (Task Id: ct:1476988200000:0:legacy_google_verify_addresses:). Reason: This task has been running too long to request a token.`

However, according to Mesos, that task had only been running for about a second (those timestamps are the start and stop time, respectively): 

`ct:1476988200000:0:legacy_google_verify_addresses: ChronosTask:legacy_google_verify_addresses  FAILED  2016-10-20T14:30:38-0400    2016-10-20T14:30:39-0400`

This is intermittent and happening to both nodes of a two-node VGM installation.  VGM and the Chronos tasks are all running in docker containers, and the Docker hosts as well as the Mesos masters are all running NTP and are in sync date-wise.  

Any ideas on how to go about further debugging?

### PR DESCRIPTION
**Pull request by jkt628**
pretty much ignore the vendor directory and focus on other commits

fixes #19 along the way

### CODE REVIEW

**Comment by nemothekid** [comment]
Thanks for your work - I see you moved multiple packages to the vendor directory - can you just outline the other changes you've made?

File: provider.go [hunk 1]
```diff
@@ -174,20 +168,9 @@ func Provide(c *gin.Context) {
...
-				}{string(state.Status), false, errTaskNotFresh.Error()})
-				return
-			}
 			// https://github.com/apache/mesos/blob/a61074586d778d432ba991701c9c4de9459db897/src/webui/master/static/js/controllers.js#L148
-			startTime := time.Unix(0, int64(task.Statuses[0].Timestamp*1000000000))
-			if time.Now().Sub(startTime) > config.MaxTaskLife {
-				log.Printf("Rejected token request from %s (Task Id: %s). Reason: %v (no status)", remoteIp, reqParams.TaskId, errTaskNotFresh)
+			if len(task.Statuses) > 0 && time.Now().Sub(time.Unix(0, int64(task.Statuses[0].Timestamp*1000000000))) > config.MaxTaskLife {
+				log.Printf("Rejected token request from %s (Task Id: %s). Reason: %v", remoteIp, reqParams.TaskId, errTaskNotFresh)
 				atomic.AddInt32(&state.Stats.Denied, 1)
```
Comment (nemothekid) [inline provider.go]: Am I missing something here? with this change, if `len(task.Statuses) == 0`, we will issue a token which is wrong.

**Comment by jkt628** [comment]
i tried to group distinct changes per commit; it is possible there is bleed over.  the commit messages should provide an overview of all the changes.

File: provider.go (hunk 1 above)
Comment (jkt628) [inline provider.go]: i tested this with STAGING applications and it works as long as the MESOS_TASK_ID is valid.

**Comment by nemothekid** [comment]
Yes, I can see that, but I don't udnerstand why you made some changes. For example in your ["	STAGING is fresh"](https://github.com/ChannelMeter/vault-gatekeeper-mesos/pull/20/commits/c590425284bd6f80ec12b9b5a3e44607c5636b44) commit, you changed how we handled tasks with 0 length task.

The problem with this Mesos isn't explicitly documented to have a task with a 0-length task status to be STAGING, so I can't say for sure that a 0-length task status is STAGING. This could cause bugs where we hand out tokens to tasks that have invalid processes or if there other conditions that cause a 0-length task status.

The previous code worked around this by waiting until a task had a non-zero status. What was the reasoning for removing this work around?

**Comment by jkt628** [comment]
my use case intercepts the docker socket to do some work on STAGING tasks.  due to the intricacies of where i operate the tasks become RUNNING longer after i've done my work.

**Comment by jkt628** [comment]
i did look for something that explicitly indicates STAGING; do you know of such?

**Pull request closed without merging**